### PR TITLE
TreeView: Filtering, Sorting, Grouping

### PR DESCRIPTION
**Pull request by falko17**
Apart from a lot of smaller improvements to the TreeWindow, this pull request adds the following functionality (accessible via buttons next to the search bar):
- **Filtering**: Both the normal tree view and the graph search can apply a combination of filters, handled by the `GraphFilter` class. Currently, this only includes toggle attributes and a toggle for edges. Filtered elements will not be shown in the TreeWindow (orphan nodes and dangling edges are "lifted" upwards).
- **Sorting**: Both the normal tree view and the graph search can sort the results by a number of attributes via the `GraphSorter` class. Currently, this includes the source name, the source line, the filename, and the type. Sorters are applied in order, and can be made descending or ascending on a case-by-case basis.
- **Grouping**: In the tree view, elements can be grouped together (i.e., put below a root node together) via the `TreeWindowGrouper` class. Currently, this is possible with the reflexion state and the element type. Only one group can be applied at a time. The hierarchy is preserved here – for details on what this means, see #680. Elements that do not actually belong to the group are rendered in italics.

All three modifiers can be combined in any desired way (e.g., filters can still be applied while grouping elements). Additionally, it should be very easy to include additional filters/sorters/groupers with this structure in place.

## Additional Changes
* A static `Icons` class has been added, so that icon glyphs can be easily reused across different contexts. See https://github.com/uni-bremen-agst/SEE/wiki/Icons for details.
* Headings for PopupMenus have been implemented.
* A bug in the MoveAction has been fixed that led to nodes not being unparented correctly in a reflexion analysis context.
* An option has been added to keep PopupMenus open even after clicking on them.
* A `DefaultDictionary` has been added that adds and returns a default value if an accessed dictionary key does not actually exist.
* A bug in the CI has been fixed which led to bad patterns not being detected correctly.
* Unity has been updated to version 2022.3.15f1.
* Additional smaller bug fixes.

### CODE REVIEW

**Review by koschke — CHANGES_REQUESTED** [review]
I have only very trivial remarks.

File: Assets/SEE/DataModel/GraphSearch/GraphFilter.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+using SEE.DataModel.DG;
+
+namespace SEE.DataModel.GraphSearch
+{
+    /// <summary>
+    /// A configurable filter for graph elements, mainly intended for use with <see cref="GraphSearch"/>.
+    /// </summary>
+    public record GraphFilter : IGraphModifier
+    {
+        // if empty, include all. otherwise, *at least one* must be present.
```
[the remark this answers is not in the document]
Comment (falko17) [reply Assets/SEE/DataModel/GraphSearch/GraphFilter.cs:12]: My bad, I forgot to delete this after writing proper documentation for this class. Deleted it now.

File: Assets/SEE/DataModel/DG/Graph.cs [hunk 2]
```diff
@@ -486,15 +486,19 @@ public void RemoveElement(GraphElement element)
         /// Returns the names of all node types of this graph
```
[the remark this answers is not in the document]
Comment (falko17) [reply Assets/SEE/DataModel/DG/Graph.cs:486]: Fixed.

File: Assets/SEE/DataModel/DG/Graph.cs [hunk 3]
```diff
@@ -486,15 +486,19 @@ public void RemoveElement(GraphElement element)
...
-            foreach (Node node in Nodes())
-            {
-                result.Add(node.Type);
-            }
-            return result;
-        }
+        internal HashSet<string> AllNodeTypes() => Nodes().Select(n => n.Type).ToHashSet();
+
+        /// <summary>
+        /// Returns the names of all edge types of this graph
```
[the remark this answers is not in the document]
Comment (falko17) [reply Assets/SEE/DataModel/DG/Graph.cs:492]: Fixed.

File: Assets/SEE/DataModel/DG/Graph.cs [hunk 4]
```diff
@@ -486,15 +486,19 @@ public void RemoveElement(GraphElement element)
...
+        /// </summary>
+        /// <returns>edge types of this graph</returns>
+        internal HashSet<string> AllEdgeTypes() => Edges().Select(e => e.Type).ToHashSet();
+
+        /// <summary>
+        /// Returns the names of all element types of this graph
```
[the remark this answers is not in the document]
Comment (falko17) [reply Assets/SEE/DataModel/DG/Graph.cs:498]: Fixed.

File: Assets/SEE/Utils/CollectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using FuzzySharp.Utils;
+
+namespace SEE.Utils
+{
+    public static class CollectionExtensions
```
[the remark this answers is not in the document]
Comment (falko17) [reply Assets/SEE/Utils/CollectionExtensions.cs:7]: Fixed.

File: Assets/SEE/DataModel/GraphSearch/GraphSorter.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            if (result == default)
+            {
+                return null;
+            }
+            else
+            {
+                return result.Descending;
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (falko17) [reply Assets/SEE/DataModel/GraphSearch/GraphSorter.cs:76]: Fixed.

File: Assets/SEE/DataModel/GraphSearch/GraphSearch.cs [hunk 7]
```diff
@@ -26,15 +27,27 @@ public class GraphSearch : IObserver<ChangeEvent>
...
+        /// The filter that is applied to the graph elements when they are searched.
+        /// </summary>
+        public GraphFilter Filter { get; } = new();
+
+        /// <summary>
+        /// The sorter that is applied to the graph elements when they are searched.
+        /// </summary>
+        public GraphSorter Sorter { get; } = new();
+
+        private IEnumerable<IGraphModifier> Modifiers => new IGraphModifier[] { Filter, Sorter };
```
[the remark this answers is not in the document]
Comment (falko17) [reply Assets/SEE/DataModel/GraphSearch/GraphSearch.cs:42]: Yes, I added it now.

File: Assets/SEE/UI/Window/TreeWindow/TreeWindowContextMenu.cs [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+                }, Icons.ArrowRotateLeft, CloseAfterClick: false)
+            };
+
+            if (Grouper.IsActive)
+            {
+                entries.Add(new PopupMenuHeading("Grouping is active!"));
+                entries.Add(new PopupMenuHeading("Items ordered by group count."));
+            }
+
+            // TODO: Any other attributes we want to sort by? Or should we just include all attributes in the future?
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEE/UI/Window/TreeWindow/TreeWindowContextMenu.cs:248]: Didn't know that was a question to me.

These are the attributes we want to sort by for the time being. We might want to include all other attributes in the future, in which case the code can be adapted. Let's wait and see whether a need for that arises at all.

I turned my review comment into a code comment and removed the TODO.

File: Assets/SEE/UI/Window/TreeWindow/TreeWindowContextMenu.cs [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+                    UpdateGroupMenuEntries();
+                }, Radio(!Grouper.IsActive), CloseAfterClick: true),
+                GroupActionFor("Reflexion State",
+                               new TreeWindowGroupAssigment<State?>(Enum.GetValues(typeof(State)).Cast<State?>()
+                                                                        .ToDictionary(keySelector: x => x,
+                                                                                      elementSelector: ReflexionStateToGroup),
+                                                                    element => element is Edge edge ? edge.State() : null)),
+                GroupActionFor("Type",
+                               new TreeWindowGroupAssigment<string>(Searcher.Graph.AllElementTypes().ToDictionary(x => x, TypeToGroup), element => element.Type)),
+                // TODO: Any other useful groups?
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEE/UI/Window/TreeWindow/TreeWindowContextMenu.cs:354]: Didn't know that was a question to me.

As above, let's wait and see whether we need additional grouping criteria. I removed the TODO and added a code comment:

```
// Here we define the criteria by which a user can group. If new grouping criteria
// arise in the future, they need to be added here.
```

File: Assets/SEE/DataModel/GraphSearch/GraphFilter.cs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        /// </summary>
+        /// <param name="elements">The elements to filter.</param>
+        /// <typeparam name="T">The type of the elements.</typeparam>
+        /// <returns>The filtered elements.</returns>
+        public IEnumerable<T> Apply<T>(IEnumerable<T> elements) where T : GraphElement
+        {
+            return elements.Where(Includes);
+        }
+
+        public bool IsActive() => !IncludeNodes || !IncludeEdges
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEE/DataModel/GraphSearch/GraphFilter.cs:67]: I added a <see cref=...> reference to the interface method implemented.

File: Assets/SEE/DataModel/GraphSearch/GraphFilter.cs [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            || IncludeToggleAttributes.Count > 0 || ExcludeToggleAttributes.Count > 0 || ExcludeElements.Count > 0;
+
+        public void Reset()
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEE/DataModel/GraphSearch/GraphFilter.cs:70]: I added a <see cref=...> reference to the interface method implemented.

File: Assets/SEE/DataModel/GraphSearch/GraphSorter.cs [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+        public bool IsActive() => SortAttributes.Count > 0;
+
+        public void Reset() => SortAttributes.Clear();
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEE/DataModel/GraphSearch/GraphSorter.cs:79]: I added a <see cref=...> reference to the interface method implemented.

File: Assets/SEE/DataModel/GraphSearch/GraphSorter.cs [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+                return null;
+            }
+            else
+            {
+                return result.Descending;
+            }
+        }
+
+
+        public bool IsActive() => SortAttributes.Count > 0;
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEE/DataModel/GraphSearch/GraphSorter.cs:77]: I added a <see cref=...> reference to the interface method implemented.

**Review by koschke — APPROVED** [review]
All fine. I resolved all remaining issues myself.

**Pull request merged**